cluster_experiments/random_splitter.py

```diff
@@ -198,3 +193,59 @@ def assign_treatment_df(
     def from_config(cls, config):
         """Creates a NonClusteredSplitter from a PowerConfig"""
         return cls(treatments=config.treatments, treatment_col=config.treatment_col)
+
+
+class StratifiedClusteredSplitter(RandomSplitter):
+    def __init__(
+        self,
+        cluster_cols: Optional[List[str]] = None,
+        treatments: Optional[List[str]] = None,
+        treatment_col: str = "treatment",
+        strata_cols: Optional[List[str]] = None,
+    ) -> None:
+        super().__init__(
+            cluster_cols=cluster_cols,
+            treatments=treatments,
+            treatment_col=treatment_col,
+        )
+        self.strata_cols = strata_cols or ["strata"]
+
+    def assign_treatment_df(self, df: pd.DataFrame) -> pd.DataFrame:
+        df = df.copy()
+        df_unique = df.loc[:, self.cluster_cols + self.strata_cols].drop_duplicates()
+
+        # check that, for a given cluster, there is only 1 strata
+        for strata_col in self.strata_cols:
+            if df_unique.groupby(self.cluster_cols)[strata_col].nunique().max() > 1:
+                raise ValueError(
+                    f"There are multiple values in {strata_col} for the same cluster item"
+                    "You cannot stratify on this column"
+                )
+
+        # random shuffling
+        random_sorted_treatments = list(np.random.permutation(self.treatments))
+
+        df_unique_shuffled = df_unique.sample(frac=1)
```

Comment:
Couple things: you could do this in line 215, in a more pipeline manner, and also do a reset_index(drop=True) after this to have less noise with indices

Reply:
done!